### PR TITLE
[MDEP-698] replace code that relies on default encoding

### PR DESCRIPTION
**Pull request by elharo**
@michael-o

### CODE REVIEW

**Review by michael-o — COMMENTED** [review]
Shouldn't we use the output encoding property from the POM and then fall back to system default when nothing is defined? Like we do for reporting and other spots.

**Comment by elharo** [comment]
Using the output encoding property makes sense. Let me figure out how to do that. 

```
 <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
```

**Comment by michael-o** [comment]
> if it's not set though, falling back to UTF-8 is more likely to be correct than the platform default.

While I agree with this, it will be inconsistent with our behavior. Please check whether UTF-8 or platform is used when the property is missing. Look in the source code of https://maven.apache.org/plugins/maven-site-plugin/site-mojo.html#outputEncoding.

**Comment by michael-o** [comment]
Why not just like [here](https://maven.apache.org/plugins/maven-site-plugin/xref/org/apache/maven/plugins/site/render/AbstractSiteRenderingMojo.html#L168) and [here](https://maven.apache.org/plugins/maven-site-plugin/xref/org/apache/maven/plugins/site/render/AbstractSiteRenderingMojo.html#L208)? Consistent usage for the user.

**Comment by elharo** [comment]
I noticed this:

```
    /**
     * Attach the classpath file to the main artifact so it can be installed and deployed.
     *
     * @since 2.0
     */
    @Parameter( defaultValue = "false" )
    private boolean attach;
```

If the classpath file we're building here is going out to Maven reppositories, it really shouldn't use the platform encoding. What's the purpose of this file? Is there a spec for it somewhere that might indicate which encoding should be used?

**Comment by michael-o** [comment]
> 
> 
> I noticed this:
> 
> ```
>     /**
>      * Attach the classpath file to the main artifact so it can be installed and deployed.
>      *
>      * @since 2.0
>      */
>     @Parameter( defaultValue = "false" )
>     private boolean attach;
> ```
> 
> If the classpath file we're building here is going out to Maven repositories, it really shouldn't use the platform encoding. What's the purpose of this file? Is there a spec for it somewhere that might indicate which encoding should be used?

Stick to the MSITE approach. If not defined, use UTF-8.

**Review by elharo — COMMENTED** [review]
done

File: src/main/java/org/apache/maven/plugins/dependency/fromDependencies/BuildClasspathMojo.java [hunk 1]
```diff
@@ -322,18 +326,20 @@ private boolean isUpdToDate( String cpString )
...
-     * @throws MojoExecutionException
+     * @param cpString the string to write into the file
      */
     private void storeClasspathFile( String cpString, File out )
         throws MojoExecutionException
     {
         // make sure the parent path exists.
         out.getParentFile().mkdirs();
+        
+        String encoding = getProject().getProperties().getProperty( "project.reporting.outputEncoding", "UTF-8" );
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/dependency/fromDependencies/BuildClasspathMojo.java:339]: This line does not make sense anymore.

File: src/main/java/org/apache/maven/plugins/dependency/fromDependencies/BuildClasspathMojo.java [hunk 2]
```diff
@@ -322,18 +326,20 @@ private boolean isUpdToDate( String cpString )
...
 
-        try ( Writer w = new BufferedWriter( new FileWriter( out ) ) )
+        try ( Writer w =
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/dependency/fromDependencies/BuildClasspathMojo.java:341]: Why the pain? `Files.newBufferedWriter(...)`?

File: src/main/java/org/apache/maven/plugins/dependency/fromDependencies/BuildClasspathMojo.java [hunk 3]
```diff
@@ -366,7 +372,10 @@ protected String readClasspathFile()
             return null;
         }
         StringBuilder sb = new StringBuilder();
-        try ( BufferedReader r = new BufferedReader( new FileReader( outputFile ) ) )
+        String encoding = outputEncoding == null ? "UTF-8" : outputEncoding;
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/dependency/fromDependencies/BuildClasspathMojo.java:375]: I'd copy the same method from MSITE to here.

**Review by michael-o — CHANGES_REQUESTED** [review]

**Comment by slachiewicz** [comment]
Please squash and rebase

File: src/main/java/org/apache/maven/plugins/dependency/fromDependencies/BuildClasspathMojo.java (hunk 1 above)
Comment (elharo) [reply src/main/java/org/apache/maven/plugins/dependency/fromDependencies/BuildClasspathMojo.java]: missed that one. Fixed.

File: src/main/java/org/apache/maven/plugins/dependency/fromDependencies/BuildClasspathMojo.java (hunk 2 above)
Comment (elharo) [reply src/main/java/org/apache/maven/plugins/dependency/fromDependencies/BuildClasspathMojo.java:342]: Old habits mostly, but looking at that it requires as much code to change types (File --> Path, String --> charset, boolean --> FileOption) that it's equally complex.

File: src/main/java/org/apache/maven/plugins/dependency/fromDependencies/BuildClasspathMojo.java (hunk 3 above)
Comment (elharo) [reply src/main/java/org/apache/maven/plugins/dependency/fromDependencies/BuildClasspathMojo.java]: Maybe just

```
        String encoding = Objects.toString( outputEncoding, "UTF-8" );
```

File: src/main/java/org/apache/maven/plugins/dependency/tree/TreeMojo.java [hunk 4]
```diff
@@ -286,7 +286,8 @@ public void execute()
 
             if ( outputFile != null )
             {
-                DependencyUtil.write( dependencyTreeString, outputFile, this.appendOutput, getLog() );
+                String encoding = project.getProperties().getProperty( "project.reporting.outputEncoding", "UTF-8" );
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/dependency/tree/TreeMojo.java:289]: Here is a leftover.

**Review by michael-o — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/maven/plugins/dependency/tree/TreeMojo.java (hunk 4 above)
Comment (elharo) [reply src/main/java/org/apache/maven/plugins/dependency/tree/TreeMojo.java]: fixed

**Review by michael-o — APPROVED** [review]

**Pull request merged**